### PR TITLE
#320. Added cryptography support.

### PR DESCRIPTION
**Pull request by tomas-knap**
Not mandatory, turned off by default, does not intercept export / import DPU nor copy pipeline.

### CODE REVIEW

**Comment by tomas-knap** [comment]
Reviewing

**Comment by tomas-knap** [comment]
Two major comments, A) is a MUST, B) is SHOULD. 

### A) Cryptography must be always optional. 

Please add one config param to config.properties which turns on/off the cryptography (turned off by default). You may also add information, that changing this value will cause that certain config.param values and existing configurations will not be properly read. So basically such value should be set up by the installer (by default turned off). 

If crypto is turned off, config params and DPU configurations MUST not be tried to be encrypted/decrypted. 

### B) Please use some standard library for encrypt/decrypt. 

For example, Jasypt has nice integration with Spring [1,2]. Integration with maven [3]. 

Using directly low level Crypto class (as you did) means that we may encounter e.g. certain byte-conversion problem when moving encrypted configuration between linux and win deployments (e.g.), or you may introduce some weakness of the algorithm due to, e.g., not properly initializing the algorithm. So using library for encrypt/decrypt is preferable. And it should be easy as shown above.

So basically using Jasypt, you:

1) Define new bean
```
<bean id="strongEncryptor"
    class="org.jasypt.encryption.pbe.StandardPBEStringEncryptor">
    <property name="algorithm">
        <value>PBEWithMD5AndTripleDES</value>
    </property>
    <property name="password">
        <value>jasypt</value> <!-- take the value from config.properties -->
    </property>
  </bean>
```


2) add to AppConfig and DPURecord classes: 

@Autowired
StandardPBEStringEncryptor crypto;

3) use crypto.encrypt(), crypto.decrypt() to encrypt/decrypt

 








[1] http://www.jasypt.org/spring3.html
[2] http://www.jasypt.org/api/jasypt/1.8/org/jasypt/encryption/pbe/StandardPBEStringEncryptor.html
[3] http://www.jasypt.org/maven.html

**Comment by tomas-knap** [comment]
Please discuss with Michal Rajniak / @skrchnavy whether B) will be done or not. I suggest to address also B), it should be work for 0.5 MD, including testing.

**Comment by eea01** [comment]
A) If cryptography.key.file is not defined in the configuration file, cryptography is turned off, which is default behavior. What else should be additionally implemented?
B) Will not be implemented at the moment.

**Comment by tomas-knap** [comment]
Ok, so what will be the behaviour when the key is not set up and method below is called? 

```
        serializedConfiguration = AppConfig.getCryptography().encrypt(serializedConfiguration);
```

Will retest and let you know

**Comment by eea01** [comment]
If cryptography key is not set, mentioned method returns unchanged input value. See [Doc Comment](https://github.com/UnifiedViews/Core/blob/41a262a1cc4376d262b76b8a3094ecf8e31274b0/commons-app/src/main/java/eu/unifiedviews/commons/util/Cryptography.java#L41).

**Comment by tomas-knap** [comment]
Yes, but this is a wrong design decision, from more reasons: 

1) Classes and methods should be used for the purpose for which they were developed. If Cryptography class is used to convert string s to s, it is wrong. 
2) It is confusing for code maintainers, as they expect 1) holds. So if they see getCrypto().encrypt(), they assume something is encrypted. 
3) Turning cryptography on/off by setting up the key is not intuitive for the user. It is the same like starting the car by releasing handbreak. 

SO please, add config parameter "cryptography.enabled" to config properties, which is false by default. If it is false, NO Crypto object is created, NO methods encrypt(), decrypt() are called.

**Comment by eea04** [comment]
Jasypt library has come CLI utilities (encrypt/decrypt values on command line) we could use and which we need. If we migrated to jasypt, we wouldn't need to develop those ourselves.

I suggest to investigate whether Jasypt implementation covers what we need in a simple enough way, and also in a way that enables reuse of OOTB jasypt CLI utilities (i.e. PBEStringEncryptor should be used for encryption)

My assumptions about the migration:
- the signature of the main class (Cryptography) will remain as is or have only cosmetic changes
- the client code will have no or only cosmetic changes
- the implementation of Cryptography class in jasypt should be simpler or at most as complex as current implementation

**Comment by eea04** [comment]
Also, i would want the crypto implementation to function as follows to make the implementation more robust:

- if encryption is turned off in the beginning and later it is turned on, and there are are some values in config properties or in DPU configuration that remained in plaintext, the implementation should either continue to function without a problem or should immediately fail (locally, not whole application). Specifically, it shouldn't "decrypt" the plaintext value to some meaningless value and the report e.g. "wrong password"

- if encryption is turned on in the beginning and later it is turned off, and there are are some values in config properties or in DPU configuration that remained encrypted, the implementation should immediately fail. Specifically, it shouldn't use the encrypted value as plaintext value and then report e.g. "wrong password".

Is such implementation possible?

**Comment by eea01** [comment]
UV can not be started if cryptography is enabled and DPU configurations in database or values in UV configuration files are not encrypted. The same is with cryptography disabled and existing encrypted data.

In order to support cryptography switching in UV, data migration tool has to be additionally implemented to support following scenario: cryptography disabled -> cryptography enabled -> cryptography key changed -> cryptography disabled.

File: commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java [hunk 1]
```diff
@@ -152,4 +175,57 @@ public AppConfig getSubConfiguration(ConfigProperty namespace) {
     public Properties getProperties() {
         return (Properties) prop.clone();
     }
+
+    private static void postprocess(Properties properties) {
```
Comment (tomas-knap) [inline commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java]: Minor: This should not be static. Is there any reason for that?

Comment (eea01) [inline commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java]: Because it is also called by static method [AppConfig.loadFrom](https://github.com/UnifiedViews/Core/blob/feature/cryptography/commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java#L94).

**Comment by tomas-knap** [comment]
Just the minor issue above, otherwise ok.

File: commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java (hunk 1 above)
Comment (tomas-knap) [inline commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java]: Yes, but this method creates new instance of AppConfig, so it could be called as: config.postprocess();

Comment (eea01) [inline commons-app/src/main/java/cz/cuni/mff/xrg/odcs/commons/app/conf/AppConfig.java]: Refactored accordingly.

**Comment by eea01** [comment]
Merged into develop.
